File: scripts/verify-module-route-logic.py
#!/usr/bin/env python3
"""Unit-style checks for module route validation (mirrors client/src/lib/moduleRoute.ts)."""
from __future__ import annotations

import re
import sys

SAFE_MODULE_ROUTE = re.compile(r"^[a-z0-9][a-z0-9/_:-]*$")


def normalize_module_route_path(path: str) -> str:
    return path.strip().lstrip("/")


def is_safe_module_route_path(path: str) -> bool:
    normalized = normalize_module_route_path(path)
    if not normalized or ".." in normalized or "//" in normalized:
        return False
    return bool(SAFE_MODULE_ROUTE.match(normalized))


def routes_array_to_record(routes):
    if not isinstance(routes, list):
        return None
    out = {}
    for entry in routes:
        if not isinstance(entry, dict):
            continue
        path = entry.get("path")
        if isinstance(path, str):
            path = path.strip().lstrip("/")
        else:
            path = ""
        element = entry.get("element")
        if path and callable(element):
            out[path] = element
    return out if out else None


def unwrap_module_export(raw):
    if callable(raw):
        return unwrap_module_export(raw())
    if isinstance(raw, dict):
        return raw
    return {}


def normalize_module_bundle(mod):
    root = unwrap_module_export(mod.get("default", mod))
    routes_raw = root.get("routes", mod.get("routes"))
    if routes_raw and not isinstance(routes_raw, list) and isinstance(routes_raw, dict):
        routes = routes_raw
    else:
        routes = routes_array_to_record(routes_raw)
    components_raw = root.get("components", mod.get("components"))
    components = components_raw if isinstance(components_raw, dict) and not isinstance(components_raw, list) else None
    return {"routes": routes, "components": components}


def main() -> int:
    failed = 0

    def assert_ok(name: str, cond: bool) -> None:
        nonlocal failed
        if not cond:
            print(f"FAIL: {name}", file=sys.stderr)
            failed = 1
        else:
            print(f"OK: {name}")

    assert_ok("accepts wfcp path", is_safe_module_route_path("shop/wfcp-module/quick-add"))
    assert_ok("accepts param segment", is_safe_module_route_path("analytics-module/:section"))
    assert_ok("accepts settings param", is_safe_module_route_path("settings/wfcp-module/:tab"))
    assert_ok("strips leading slash", is_safe_module_route_path("/shop/wfcp-module/quick-add"))
    assert_ok("rejects traversal", not is_safe_module_route_path("../evil"))
    assert_ok("rejects double slash", not is_safe_module_route_path("shop//evil"))

    def page():
        return None

    from_array = normalize_module_bundle(
        {
            "routes": [
                {"path": "settings/shop/basalam-module", "element": page},
                {"path": "/settings/shop/basalam-module/payments", "element": page},
            ],
        }
    )
    assert_ok(
        "array routes normalize",
        from_array["routes"].get("settings/shop/basalam-module") is page
        and from_array["routes"].get("settings/shop/basalam-module/payments") is page,
    )

    from_fn = normalize_module_bundle({"default": lambda: {"routes": {"bots/bale": page}}})
    assert_ok("function export unwrap", from_fn["routes"].get("bots/bale") is page)

    return failed


if __name__ == "__main__":
    sys.exit(main())

File: scripts/verify-module-route-logic.mjs
#!/usr/bin/env node
/**
 * Unit-style checks for module route validation (mirrors client/src/lib/moduleRoute.ts).
 */
const SAFE_MODULE_ROUTE = /^[a-z0-9][a-z0-9/_:-]*$/

function normalizeModuleRoutePath(path) {
  return path.trim().replace(/^\/+/, '')
}

function isSafeModuleRoutePath(path) {
  const normalized = normalizeModuleRoutePath(path)
  if (!normalized || normalized.includes('..') || normalized.includes('//')) {
    return false
  }
  return SAFE_MODULE_ROUTE.test(normalized)
}

function routesArrayToRecord(routes) {
  if (!Array.isArray(routes)) return undefined
  const out = {}
  for (const entry of routes) {
    const path = typeof entry?.path === 'string' ? entry.path.trim().replace(/^\/+/, '') : ''
    if (path && typeof entry.element === 'function') {
      out[path] = entry.element
    }
  }
  return Object.keys(out).length > 0 ? out : undefined
}

function unwrapModuleExport(raw) {
  if (typeof raw === 'function') {
    return unwrapModuleExport(raw())
  }
  if (raw && typeof raw === 'object') {
    return raw
  }
  return {}
}

function normalizeModuleBundle(mod) {
  const root = unwrapModuleExport(mod.default ?? mod)
  const routesRaw = root.routes ?? mod.routes
  let routes
  if (routesRaw && !Array.isArray(routesRaw) && typeof routesRaw === 'object') {
    routes = routesRaw
  } else {
    routes = routesArrayToRecord(routesRaw)
  }
  const componentsRaw = root.components ?? mod.components
  const components =
    componentsRaw && typeof componentsRaw === 'object' && !Array.isArray(componentsRaw)
      ? componentsRaw
      : undefined
  return { routes, components }
}

let failed = 0

function assert(name, cond) {
  if (!cond) {
    console.error(`FAIL: ${name}`)
    failed = 1
  } else {
    console.log(`OK: ${name}`)
  }
}

assert('accepts wfcp path', isSafeModuleRoutePath('shop/wfcp-module/quick-add'))
assert('accepts param segment', isSafeModuleRoutePath('analytics-module/:section'))
assert('accepts settings param', isSafeModuleRoutePath('settings/wfcp-module/:tab'))
assert('strips leading slash', isSafeModuleRoutePath('/shop/wfcp-module/quick-add'))
assert('rejects traversal', !isSafeModuleRoutePath('../evil'))
assert('rejects double slash', !isSafeModuleRoutePath('shop//evil'))

const Page = () => null
const fromArray = normalizeModuleBundle({
  routes: [
    { path: 'settings/shop/basalam-module', element: Page },
    { path: '/settings/shop/basalam-module/payments', element: Page },
  ],
})
assert(
  'array routes normalize',
  fromArray.routes?.['settings/shop/basalam-module'] === Page &&
    fromArray.routes?.['settings/shop/basalam-module/payments'] === Page,
)

const fromFn = normalizeModuleBundle({
  default: () => ({ routes: { 'bots/bale': Page } }),
})
assert('function export unwrap', fromFn.routes?.['bots/bale'] === Page)

function stripModulePathSegment(path) {
  return normalizeModuleRoutePath(path).replace(/-module(?=\/|$)/g, '')
}

function resolveBundleRoute(routes, routePath) {
  if (!routes) return null
  const normalized = normalizeModuleRoutePath(routePath)
  if (routes[normalized]) return routes[normalized]
  if (routes[routePath]) return routes[routePath]
  const stripped = stripModulePathSegment(normalized)
  for (const [key, comp] of Object.entries(routes)) {
    if (normalizeModuleRoutePath(key) === normalized) return comp
    if (stripModulePathSegment(key) === stripped) return comp
  }
  return null
}

const legacyRoutes = {
  'analytics-module/:section': Page,
  'shop/wfcp-module/quick-add': Page,
}
assert(
  'legacy analytics alias',
  resolveBundleRoute(legacyRoutes, 'analytics/:section') === Page,
)
assert(
  'legacy wfcp alias',
  resolveBundleRoute(legacyRoutes, 'shop/wfcp/quick-add') === Page,
)
assert(
  'clean path exact',
  resolveBundleRoute({ 'settings/shop/snapppay': Page }, 'settings/shop/snapppay') === Page,
)
assert('strip module segment', stripModulePathSegment('shop/wfcp-module/quick-add') === 'shop/wfcp/quick-add')

if (failed) {
  process.exit(1)
}
